arch: custom `_current` implementation

ycsin opened the pull request:
Allow architecture to implement their way to get the current thread for the current implementation.

___

# Benchmarks

<details>
  <summary>scheduler microbenchmark</summary>

  Based on the upstream [scheduler microbenchmark](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/benchmarks/sched), split (and cropped) for better readability

  ## unpend, ready, switch, pend 
  ![boxplot: scheduler - each operation](https://github.com/user-attachments/assets/932afb4f-303c-4089-974e-c5e100ebb909)

  ## tot
  ![boxplot: scheduler - total](https://github.com/user-attachments/assets/69844679-5613-4451-84f6-8be6c77f2a50)

  ## avg
  ![boxplot: scheduler - average](https://github.com/user-attachments/assets/ec2ce152-bf00-4ded-be46-152dbe48fb48)

</details>

<details>
  <summary>latency measurements</summary>

  Based on the upstream [latency measurements benchmark](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/benchmarks/latency_measure)

  ## ISR back to interrupted thread
  ![ISR back to interrupted thread](https://github.com/user-attachments/assets/ea8411c2-ac7f-4c6b-9849-b64e6ec6927e)

  ## ISR back to interrupted thread (without 99.9th)

  For better readability:

  ![ISR back to interrupted thread - zoomed](https://github.com/user-attachments/assets/d611730f-7645-4946-ba1e-0b226ed60cdc)

  ## ISR to another thread
  ![ISR to another thread](https://github.com/user-attachments/assets/ff2b6dae-c205-439a-a705-bd128dda06c2)

  ## ISR to another thread (without 99.9th)

  For better readability:

  ![ISR to another thread - zoomed](https://github.com/user-attachments/assets/6ea63c84-f4a2-49c6-9d95-958f396ed991)

  ## Other tests

  ### unit in ns
  ![Other tests (ns)](https://github.com/user-attachments/assets/1c8e71ba-b3e2-485a-acd6-abbf32a92c1e)

  ### unit in cycles
  ![Other tests (cycles)](https://github.com/user-attachments/assets/6012734d-ca0b-4001-8d88-4093737faf18)

</details>

<details>
  <summary>kernel Object Performance (sys_kernel)</summary>

  Based on the upstream [sys_kernel microbenchmark](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/benchmarks/sys_kernel), split for better readability

  ## Semaphore, LIFO, FIFO, Stacks
  ![Semaphore, LIFO, FIFO, Stacks result](https://github.com/user-attachments/assets/29f441bc-a465-47ea-b6b2-90fca3591d44)

  ## Memslab
  ![Memslabresult](https://github.com/user-attachments/assets/35fbf6a3-ad10-4dd0-80bf-ab989d4b0806)

</details>

ycsin committed:
arch: add new interfaces to set/get the current thread of current CPU

Add the following arch-specific APIs:
- arch_curr_thread()
- arch_set_curr_thread()

which allow SMP architectures to implement a faster "get current
thread pointer" than the default provided by the kernel. The 'set'
function is required for the 'get' to work, more on that later.

When `CONFIG_ARCH_HAS_CUSTOM_CURRENT_IMPL` is selected, calls to
`_current` & `k_sched_current_thread_query()` will be redirected to
`arch_curr_thread()`, which ideally should translate into a single
instruction read, avoiding the current
"lock > read CPU > read current thread > unlock" path in SMP
architectures and thus greatly improves the read performance.

However, since the kernel relies on a copy of the "current thread"s on
every CPU for certain operations (i.e. to compare the priority of the
currently scheduled thread on another CPU to determine if IPI should be
sent), we can't eliminate the copy of "current thread" (`current`) from
the `struct _cpu` and therefore the kernel now has to invoke
`arch_set_curr_thread()` in addition to what it has been doing. This
means that it will take slightly longer (most likely one instruction
write) to change the current thread pointer on the current
CPU.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
arch: riscv: implement ARCH_HAS_CUSTOM_CURRENT_IMPL with GP

Implement `arch_curr_thread()` & `arch_set_curr_thread()`
with the global pointer (GP) register.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented on kernel/thread.c:
```diff
@@ -946,8 +946,8 @@ void z_thread_mark_switched_out(void)
 #ifdef CONFIG_TRACING
 #ifdef CONFIG_THREAD_LOCAL_STORAGE
 	/* Dummy thread won't have TLS set up to run arbitrary code */
```
what if we set up TLS for dummy threads?

andyross replied:
Dummy threads never run, so will never access TLS.  The idea is just a trick for re-using the context switch code for initial startup, you need some "thread" to "save" to when you want to start running your first thread but are still on a boot stack somewhere.

ycsin commented on kernel/sched.c:
```diff
@@ -1242,6 +1242,9 @@ static inline void z_vrfy_k_wakeup(k_tid_t thread)
+#ifdef CONFIG_ARCH_HAS_CUSTOM_CURRENT_IMPL
+	return arch_current_thread();
+#else
 #ifdef CONFIG_SMP
```
probably can just call `arch_current_thread()` here, and use the current implementation as the default when `CONFIG_ARCH_HAS_CUSTOM_CURRENT_IMPL=n`

ycsin commented:
cc @srv-meta

cfriedt reviewed (dismissed):
The GP reg - makes a lot of sense for RISC-V. Thank you for the figures @ycsin.

The only thing that stuck out to me as odd was that suspending and resuming a thread seemed to take slightly longer in some graphs down below.

Please take a look at this @peter-mitsis, @andyross, @npitre @fkokosinski

nashif commented on arch/riscv/Kconfig:
```diff
@@ -27,6 +27,14 @@ config RISCV_GP
 	  using GP relative addressing.
 
+config RISCV_CURRENT_VIA_GP
+	bool "Store current thread pointer into the GP register"
```
how about `Global Pointer (GP) register`.
Makes it easier to read and understand what GP is...

nashif replied:
probably also in the kconfig itself, spell it out.

ycsin replied:
I updated the prompt and the help section to spell out the 'global pointer (GP)' but kept the Kconfig name to be consistent with [`CONFIG_RISCV_GP`](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_RISCV_GP)

akabaev commented:
The only dummy thread that should matter here is the one that Zephyr uses during startup before switching to main thread, no? Should zephyr consider to possibly
- run on main thread from the start and eliminate the need for dummy thread entirely
- maybe allocate/reserve TLS for main thread early, use it from dummy thread and then inherit it from main during the initial switch?

Then TLS can be enabled for _current retrieval.

npitre commented:
Did you look at the compiled code's assembly out put to see what could
make a difference? Especially with `CONFIG_CURRENT_THREAD_USE_TLS=y` it
should be about equivalent already. And `CONFIG_RISCV_GP=y` (which conflicts
with this PR) would itself add some additional performance gain.

ycsin commented:
> Did you look at the compiled code's assembly out put to see what could make a difference? Especially with `CONFIG_CURRENT_THREAD_USE_TLS=y` it should be about equivalent already. And `CONFIG_RISCV_GP=y` (which conflicts with this PR) would itself add some additional performance gain.

I do not have kernel benchmark with `CONFIG_CURRENT_THREAD_USE_TLS` & `CONFIG_RISCV_GP` (I can probably run them tomorrow), they weren't a clear win in our internal application benchmarks.

The `CONFIG_CURRENT_THREAD_USE_TLS` only affects the `k_current_get()` calls and is per-thread, which is different from `_current` (per-cpu) and isn't used much in the kernel/scheduler

npitre commented:
> The `CONFIG_CURRENT_THREAD_USE_TLS` only affects the `k_current_get()`
> calls and is per-thread, which is different from `_current` (per-cpu) 
> and isn't used much in the kernel/scheduler

`_current` is also per-thread. How can't it be? After all it also returns
the current thread. And both methods are conceptually per-CPU as there is only
one thread at a time on each CPU.

The actual difference is that `k_current_get()` is usable from user mode.
But there is nothing fundamental preventing `_current` from leveraging the
TLS as well.

ycsin commented:
> `_current` is also per-thread. How can't it be? After all it also returns the current thread. And both methods are conceptually per-CPU as there is only one thread at a time on each CPU.
> 
> The actual difference is that `k_current_get()` is usable from user mode. But there is nothing fundamental preventing `_current` from leveraging the TLS as well.

I could be wrong, but the idea behind `_current` & `k_current_get()` seems to be fundamentally different to me - I think they can be better documented to prevent ambiguity and incorrect usage.

`_current`(or `_current_cpu->current`) returns the thread scheduled on the current CPU. When the scheduler decides that the current CPU should execute another thread and update the `_kernel.cpus[i].current`, `_current` will then point to the next thread, while `k_current_get()` will still point to the current (or rather, previous, from the CPU's perspective) thread.

Utilizing TLS in `_current` would require the scheduler to update the `z_tls_current` in the `thread A`'s TLS stack to point at `thread B` when swapping from A to B. The next time a CPU runs `thread A` and calls `k_current_get()`, `thread B`'s pointer would be return, which is wrong. 

In that sense, `k_current_get()` is per-thread, and `_current` is per-cpu and can not be per-thread.

ycsin commented:
# CONFIG_RISCV_GP=y

## latency_measure
![newplot](https://github.com/user-attachments/assets/56a658ff-6f7d-4f6d-ad1d-9d688499a7b1)
![newplot (1)](https://github.com/user-attachments/assets/88d650c3-d30c-45d1-ac5b-0891e7c5c603)
![newplot (2)](https://github.com/user-attachments/assets/979f45b8-0c98-469a-834c-1c246e87215a)
![newplot (3)](https://github.com/user-attachments/assets/9a9265bc-2daa-4e7c-8bd5-7b6b1b5b7627)
![newplot (4)](https://github.com/user-attachments/assets/38240dcf-678e-4653-8818-dd640e783af0)
![newplot (5)](https://github.com/user-attachments/assets/178dd8ca-0cc3-4ffa-a14f-621cf48e8a3a)

# scheduler
![newplot (6)](https://github.com/user-attachments/assets/83eb0102-587b-4140-9da6-f4419b27a14d)
![newplot (7)](https://github.com/user-attachments/assets/0cb75741-187c-4a80-a481-d8acd5ce3374)
![newplot (8)](https://github.com/user-attachments/assets/91954f7a-0383-46d0-88f3-791ebb47eed1)

# sys_kernel

## memslab
![newplot (9)](https://github.com/user-attachments/assets/7890e9a2-e560-42fd-bf7c-7d913659d7c2)

## others
![newplot (10)](https://github.com/user-attachments/assets/555bd2f5-329b-4927-8235-9b3148441680)

# CONFIG_THREAD_LOCAL_STORAGE=y

## latency_measure
![newplot](https://github.com/user-attachments/assets/cfa363ae-377a-4047-8894-4af6524508d3)
![newplot (1)](https://github.com/user-attachments/assets/7a0f13f8-4412-4d8d-8bae-e1514728bc65)
![newplot (2)](https://github.com/user-attachments/assets/9370d057-b6e2-4609-b631-41ae561b078e)
![newplot (3)](https://github.com/user-attachments/assets/2ddee9ff-0945-441e-87e4-9f85dab6b4e9)
![newplot (4)](https://github.com/user-attachments/assets/9dbd95c6-05f1-4243-be99-007e72afb057)
![newplot (5)](https://github.com/user-attachments/assets/657537cd-9cb5-433f-90f7-ab339c59d4a7)

# scheduler
![newplot (6)](https://github.com/user-attachments/assets/e5e654c3-9b71-48e8-a022-f4aca1508d92)
![newplot (7)](https://github.com/user-attachments/assets/9d6240af-520e-49c9-923c-9a7895c6d55c)
![newplot (8)](https://github.com/user-attachments/assets/82da5616-904a-484b-b7e5-1e1515905173)

## memslab
![newplot (9)](https://github.com/user-attachments/assets/cf46bc16-5c45-4c5f-836c-e10b05f55b05)

## others
![newplot (10)](https://github.com/user-attachments/assets/35f5aa7e-d86c-49d7-baad-5836a20587f7)

andyross reviewed (commented):
First big question: is the GP register "privileged" or "secure"?  Or can userspace code write to it?  I found some blog posts talking about using it for linker-visible symbol lookup in the toolchain, which makes me think it's probably not a kernel-only register.

If it's not safe from userspace threads, you need some code on kernel entry (syscalls and interrupt/exceptions) to set it correctly in case the untrusted code has messed with it.  Or else flag the feature with kconfig depdencies so that it isn't possible to combine with USERSPACE=y.

Beyond that I think this looks clean to me.  I tend to agree with the comment that having a "fallback" arch_current_thread() that code can use everywhere when the arch doesn't define this would be better than #if'ing on the kconfig.

peter-mitsis commented:
For what it is worth, I like this approach.

ycsin commented on kernel/include/kswap.h:
```diff
@@ -28,6 +28,15 @@ extern struct k_spinlock _sched_spinlock;
  */
 void z_smp_release_global_lock(struct k_thread *thread);
 
+/* Update the currently scheduled thread for the current CPU 8 */
```
```suggestion
/* Update the currently scheduled thread for the current CPU */
```

ycsin commented:
> First big question: is the GP register "privileged" or "secure"? Or can userspace code write to it? I found some blog posts talking about using it for linker-visible symbol lookup in the toolchain, which makes me think it's probably not a kernel-only register.

I tried to read/write something to the `gp` reg from a userspace thread and nothing happened, in contrast, reading the mscratch csr caused Illegal Instruction error, so I guess this means that it is not secure?

Then this leads me to wonder if there's a security issue with the `RISCV_GP` implementation - what if a rogue user thread modified the `gp` to some random offset? Unlike the `tp` reg, we are not resetting the `gp` reg in the `_isr_wrapper`, any idea? cc @fkokosinski @cwshu

npitre commented:
Obviously, if gp is relied upon by privileged code, it must be sanitized
when coming back from untrusted code, just like tp currently is.

cwshu commented:
GP register is just a general-purpose register, so CPU in each privileged could access it.
I think user thread could use GP if GP-relative addressing is not enabled when toolchain compiles (links) Zephyr ELF file.

With `CONFIG_RISCV_GP`, it is used by GP-relative addressing, so linker relaxation will optimize instruction to use GP register. In this case, we can't overwrite in the Zephyr user thread.
When we disable `CONFIG_RISCV_GP`, RISC-V linker script doesn't provide `__global_pointer$` symbol. It may or may not disable GP-relative addressing (I am not sure, and it may depend on toolchain). However, if Zephyr CI doesn't break, I think it could correctly disable GP-relative addressing at least in Zephyr toolchain.
If GP-relative addressing is not automatically disabled, we can use `-mno-relax` linker option to disable it manually. However, this option will increase code size significantly, because it not only disables GP-relative addressing but also all linker relaxations.

In conclusion, if Zephyr CI could pass, I think it is okay to overwrite GP via user thread.

npitre reviewed (commented):
Suggestion: unless there is a proliferating need to play with register access, 
you may consider dispensing with riscv/reg.h and do the inline asm in
arch_current_thread() directly instead. In particular, this will give you 
the freedom to _not_ put a "volatile" on the register read in order to let
the compiler perform code simplification optimization. On the other hand, the
register write does need a volatile to force execution. Maybe there is already
an implicit volatile on inline asm when no output constraints are listed
but it is best to be explicit.

Yet a better solution would consist in what the GCC manual calls a "Global
Register Variable". It would go like this:

```
register struct k_thread *__arch_current_thread asm ("gp");

#define arch_current_thread() __arch_current_thread
#define arch_set_current_thread((thread) __arch_current_thread = (thread)
```

With luck, the compiler would be able to refer to it without actually having
to copy the value to another register.

ycsin commented:
> I tend to agree with the comment that having a "fallback" arch_current_thread() that code can use everywhere when the arch doesn't define this would be better than #if'ing on the kconfig.

PTAL #81153

ycsin commented:
> With luck, the compiler would be able to refer to it without actually having
> to copy the value to another register.

Interesting, applied the suggestion to the patch

npitre commented:
Well... if you recompile using `CONFIG_OUTPUT_DISASSEMBLY=y` you'll 
notice that currently every `_current` reference is turned into an 
out-of-line call to `z_impl_k_sched_current_thread_query`. This is way 
suboptimal.

Redefining `_current` to use `arch_current()` directly would be one way 
to get this feature's full potential, and most probably make the 
compiled code smaller too.

And for test coverage purpose this should default to y.

ycsin commented:
> I tend to agree with the comment that having a "fallback" arch_current_thread() that code can use everywhere when the arch doesn't define this would be better than #if'ing on the kconfig.

To do that I have to create a common `arch_inlines.h` as the fallback, but I'm not sure if that's acceptable, PTAL @ #81153

ycsin committed:
arch: deprecate `_current`

`_current` is now functionally equals to `arch_curr_thread()`, remove
its usage in-tree and deprecate it instead of removing it outright,
as it has been with us since forever.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
Updated the release & migration notes to push this closer to the finish line.

> Well... if you recompile using `CONFIG_OUTPUT_DISASSEMBLY=y` you'll notice that currently every `_current` reference is turned into an out-of-line call to `z_impl_k_sched_current_thread_query`. This is way suboptimal. Redefining `_current` to use `arch_current()` directly would be one way to get this feature's full potential, and most probably make the compiled code smaller too.

You are absolutely right, applied the suggestion, thanks.

> And for test coverage purpose this should default to y.

Undecided about this yet, would this affect the `CONFIG_RISCV_GP`?

___

I renamed the API from `arch_current_thread` to `arch_curr_thread` to look more consistent with `arch_curr_cpu`

___

RFC: I wonder if I should name it as `arch_curr_cpu_thread` instead, to be more distinct from `k_current_get`?

`k_current_get` is a little tricky, but it should be fine as long as it is not used in the scheduler:
- When `CONFIG_CURRENT_THREAD_USE_TLS=n`, it is equivalent to `_current` / `k_sched_current_thread_query()`, which is implemented in a per-CPU global variable, and return thread scheduled on the current CPU, this variable can be modified in the scheduler
- Otherwise, it returns the `z_tls_current`, which depends on whatever thread that calls it, and cannot change after thread entry.

nashif commented on include/zephyr/arch/common/arch_inlines.h:
```diff
@@ -0,0 +1,45 @@
+#include <zephyr/kernel_structs.h>
+
+#ifndef CONFIG_ARCH_HAS_CUSTOM_CURRENT_IMPL
+static ALWAYS_INLINE struct k_thread *arch_curr_thread(void)
```
why not spell it out, are we trying to save on charachters here? `arch_current_thread()`

ycsin replied:
So it looks more consistent with `arch_curr_cpu()` in the [`arch_inlines.h`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/riscv/arch_inlines.h)

nashif replied:
that is something we can always change to also use `current`.

ycsin replied:
The latest version uses `arch_current_thread()`

nashif commented on include/zephyr/arch/common/arch_inlines.h:
```diff
@@ -0,0 +1,45 @@
+	return ret;
+}
+
+static ALWAYS_INLINE void arch_set_curr_thread(struct k_thread *thread)
```
`arch_current_thread_set()`?

ycsin replied:
Updated to `arch_current_thread_set()`

andyross commented on include/zephyr/arch/riscv/arch_inlines.h:
```diff
@@ -26,6 +27,17 @@ static ALWAYS_INLINE _cpu_t *arch_curr_cpu(void)
 }
 
+#ifdef CONFIG_RISCV_CURRENT_VIA_GP
+register struct k_thread *__arch_current_thread __asm__("gp");
```
Style: while asm+register variables are a really useful trick for controlling marshalling behavior around inline assembly, I think here it might be more confusing than valuable and a more traditional asm block which reads the value directly might be clearer.

andyross commented on arch/riscv/Kconfig:
```diff
@@ -28,6 +28,16 @@ config RISCV_GP
 
+config RISCV_CURRENT_VIA_GP
+	bool "Store current thread into the global pointer (GP) register"
+	depends on !RISCV_GP && !USERSPACE
```
Should document somewhere (maybe just right here in the help string) why USERSPACE is disallowed, and maybe even open an issue demanding that proper entry/exit protection for GP be implemented.

ycsin replied:
Created #81843

andyross reviewed (approved):
No reason not to +1, this all looks clean.  One note on some better docs so our grandchildren understand what the deal is with USERSPACE and the GP register.

Aesthetically I'm a little mixed on the last patch, though.  First because I'm not a big fan of giant refactoring, second because 13 bytes of extra typing starts to get beyond my annoyance threshold.  But mostly it's just that on the uniprocessor systems that are Zephyr's ancestral homeland, "_current" really is just a global struct and it's helpful to remember that.

cfriedt reviewed (approved):
🚀

npitre commented:
I just saw the deprecation part after being merged in main.

This PR didn't include that part initially. Or did it?

What is the rationale for that part?

Personally I'm rather against this. Unless there is an actual downside,
the core of Zephyr always used `_current` and it was fine. It is quite
prevalent as well and the alternative is proving rather verbose.
Furthermore, as a concept, the "current thread" is not something that is
necessarily architecture specific. Having a centrally defined abstraction
also helps with instrumentation when such need comes around, etc. etc.

Again, is there a rationale for this? Otherwise I propose reverting
commit b1def7145fd1.

ycsin commented:
> This PR didn't include that part initially. Or did it?
> 
> What is the rationale for that part?

Yeah, it was added later, since `_current` functions the same as `arch_current_thread()`, I thought it was good to clean that up.

If we revert b1def7145fd158e7cec50761fece73b1d4835a2e, it would just mean that we are going to have 3 functions to get the current thread:
- `_current` - the thread scheduled on the current CPU
- `arch_current_thread()` - interchangeably with `_current`
- `k_current_get()` - the thread from where this function is called

npitre commented:
> - `_current` - the thread scheduled on the current CPU

... to be used in the kernel's internal code.

> - `arch_current_thread()` - interchangeably with `_current`

Semantically they would be different. We traditionally used `arch_*` for
architecture specific implementations. In this case it doesn't necessarily
have to be an architecture specific implementation. Probably the fallback
case would be better not be called `arch_current_thread()` for that reason
and avoid possible confusion.

> - `k_current_get()` - the thread from where this function is called

That exists mainly as a syscall for user threads. In fact, with the gp-based
implementation on RISC-V, the syscall could be dispensed with too. But in
theory, those 3 variants could potentially be different in some ways.
